File: backend/app/routers/request_hubs.py
"""Request Hubs API router."""

from typing import List
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, status, Query
from sqlalchemy import select, update, delete as sql_delete
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.orm import selectinload

from app.db.session import get_session
from app.models.request_hub import RequestHub, RequestHubTab
from app.schemas.request_hubs import (
    RequestHubCreate,
    RequestHubRead,
    RequestHubReadWithTabs,
    RequestHubUpdate,
    RequestHubTabCreate,
    RequestHubTabRead,
    RequestHubTabUpdate,
    TabsReorderRequest,
)

router = APIRouter()


# ============================================================================
# Request Hubs Endpoints
# ============================================================================

@router.get("/workspaces/{workspace_id}/request-hubs", response_model=List[RequestHubRead])
async def list_request_hubs(
    workspace_id: UUID,
    user_id: UUID = Query(..., description="User ID from auth token"),
    include_inactive: bool = False,
    session: AsyncSession = Depends(get_session),
):
    """List all request hubs in a workspace."""
    stmt = select(RequestHub).where(RequestHub.workspace_id == workspace_id)
    
    if not include_inactive:
        stmt = stmt.where(RequestHub.is_active == True)
    
    stmt = stmt.order_by(RequestHub.created_at.desc())
    
    result = await session.execute(stmt)
    hubs = result.scalars().all()
    
    return hubs


@router.get("/workspaces/{workspace_id}/request-hubs/{hub_id}", response_model=RequestHubReadWithTabs)
async def get_request_hub(
    workspace_id: UUID,
    hub_id: UUID,
    user_id: UUID = Query(..., description="User ID from auth token"),
    session: AsyncSession = Depends(get_session),
):
    """Get a specific request hub with its tabs."""
    stmt = (
        select(RequestHub)
        .options(selectinload(RequestHub.tabs))
        .where(
            RequestHub.id == hub_id,
            RequestHub.workspace_id == workspace_id
        )
    )
    
    result = await session.execute(stmt)
    hub = result.scalar_one_or_none()
    
    if not hub:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Request hub {hub_id} not found in workspace {workspace_id}"
        )
    
    return hub


@router.get("/workspaces/{workspace_id}/request-hubs/by-slug/{slug}", response_model=RequestHubReadWithTabs)
async def get_request_hub_by_slug(
    workspace_id: UUID,
    slug: str,
    user_id: UUID = Query(..., description="User ID from auth token"),
    session: AsyncSession = Depends(get_session),
):
    """Get a request hub by its slug."""
    stmt = (
        select(RequestHub)
        .options(selectinload(RequestHub.tabs))
        .where(
            RequestHub.slug == slug,
            RequestHub.workspace_id == workspace_id
        )
    )
    
    result = await session.execute(stmt)
    hub = result.scalar_one_or_none()
    
    if not hub:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Request hub with slug '{slug}' not found in workspace {workspace_id}"
        )
    
    return hub


@router.post("/workspaces/{workspace_id}/request-hubs", response_model=RequestHubReadWithTabs, status_code=status.HTTP_201_CREATED)
async def create_request_hub(
    workspace_id: UUID,
    hub_data: RequestHubCreate,
    user_id: UUID = Query(..., description="User ID from auth token"),
    session: AsyncSession = Depends(get_session),
):
    """Create a new request hub."""
    # Verify workspace_id matches
    if hub_data.workspace_id != workspace_id:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Workspace ID in path must match workspace_id in request body"
        )
    
    # Check for duplicate slug
    stmt = select(RequestHub).where(
        RequestHub.workspace_id == workspace_id,
        RequestHub.slug == hub_data.slug
    )
    result = await session.execute(stmt)
    existing_hub = result.scalar_one_or_none()
    
    if existing_hub:
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail=f"Request hub with slug '{hub_data.slug}' already exists in this workspace"
        )
    
    # Create hub
    new_hub = RequestHub(
        **hub_data.model_dump(),
        created_by=user_id
    )
    
    session.add(new_hub)
    await session.commit()
    await session.refresh(new_hub, ["tabs"])
    
    return new_hub


@router.patch("/workspaces/{workspace_id}/request-hubs/{hub_id}", response_model=RequestHubReadWithTabs)
async def update_request_hub(
    workspace_id: UUID,
    hub_id: UUID,
    hub_data: RequestHubUpdate,
    user_id: UUID = Query(..., description="User ID from auth token"),
    session: AsyncSession = Depends(get_session),
):
    """Update a request hub."""
    # Get existing hub
    stmt = (
        select(RequestHub)
        .options(selectinload(RequestHub.tabs))
        .where(
            RequestHub.id == hub_id,
            RequestHub.workspace_id == workspace_id
        )
    )
    result = await session.execute(stmt)
    hub = result.scalar_one_or_none()
    
    if not hub:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Request hub {hub_id} not found in workspace {workspace_id}"
        )
    
    # Check for slug conflicts if updating slug
    if hub_data.slug and hub_data.slug != hub.slug:
        stmt = select(RequestHub).where(
            RequestHub.workspace_id == workspace_id,
            RequestHub.slug == hub_data.slug,
            RequestHub.id != hub_id
        )
        result = await session.execute(stmt)
        existing_hub = result.scalar_one_or_none()
        
        if existing_hub:
            raise HTTPException(
                status_code=status.HTTP_409_CONFLICT,
                detail=f"Request hub with slug '{hub_data.slug}' already exists in this workspace"
            )
    
    # Update fields
    update_data = hub_data.model_dump(exclude_unset=True)
    for field, value in update_data.items():
        setattr(hub, field, value)
    
    await session.commit()
    await session.refresh(hub, ["tabs"])
    
    return hub


@router.delete("/workspaces/{workspace_id}/request-hubs/{hub_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_request_hub(
    workspace_id: UUID,
    hub_id: UUID,
    user_id: UUID = Query(..., description="User ID from auth token"),
    session: AsyncSession = Depends(get_session),
):
    """Delete a request hub."""
    stmt = select(RequestHub).where(
        RequestHub.id == hub_id,
        RequestHub.workspace_id == workspace_id
    )
    result = await session.execute(stmt)
    hub = result.scalar_one_or_none()
    
    if not hub:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Request hub {hub_id} not found in workspace {workspace_id}"
        )
    
    await session.delete(hub)
    await session.commit()


# ============================================================================
# Request Hub Tabs Endpoints
# ============================================================================

@router.get("/workspaces/{workspace_id}/request-hubs/{hub_id}/tabs", response_model=List[RequestHubTabRead])
async def list_request_hub_tabs(
    workspace_id: UUID,
    hub_id: UUID,
    user_id: UUID = Query(..., description="User ID from auth token"),
    include_hidden: bool = False,
    session: AsyncSession = Depends(get_session),
):
    """List all tabs for a request hub."""
    # Verify hub exists and belongs to workspace
    stmt = select(RequestHub).where(
        RequestHub.id == hub_id,
        RequestHub.workspace_id == workspace_id
    )
    result = await session.execute(stmt)
    hub = result.scalar_one_or_none()
    
    if not hub:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Request hub {hub_id} not found in workspace {workspace_id}"
        )
    
    # Get tabs
    stmt = select(RequestHubTab).where(RequestHubTab.hub_id == hub_id)
    
    if not include_hidden:
        stmt = stmt.where(RequestHubTab.is_visible == True)
    
    stmt = stmt.order_by(RequestHubTab.position)
    
    result = await session.execute(stmt)
    tabs = result.scalars().all()
    
    return tabs


@router.post("/workspaces/{workspace_id}/request-hubs/{hub_id}/tabs", response_model=RequestHubTabRead, status_code=status.HTTP_201_CREATED)
async def create_request_hub_tab(
    workspace_id: UUID,
    hub_id: UUID,
    tab_data: RequestHubTabCreate,
    user_id: UUID = Query(..., description="User ID from auth token"),
    session: AsyncSession = Depends(get_session),
):
    """Create a new tab in a request hub."""
    # Verify hub exists and belongs to workspace
    stmt = select(RequestHub).where(
        RequestHub.id == hub_id,
        RequestHub.workspace_id == workspace_id
    )
    result = await session.execute(stmt)
    hub = result.scalar_one_or_none()
    
    if not hub:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Request hub {hub_id} not found in workspace {workspace_id}"
        )
    
    # Check for duplicate slug
    stmt = select(RequestHubTab).where(
        RequestHubTab.hub_id == hub_id,
        RequestHubTab.slug == tab_data.slug
    )
    result = await session.execute(stmt)
    existing_tab = result.scalar_one_or_none()
    
    if existing_tab:
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail=f"Tab with slug '{tab_data.slug}' already exists in this hub"
        )
    
    # Create tab
    new_tab = RequestHubTab(
        **tab_data.model_dump(),
        hub_id=hub_id
    )
    
    session.add(new_tab)
    await session.commit()
    await session.refresh(new_tab)
    
    return new_tab


@router.patch("/workspaces/{workspace_id}/request-hubs/{hub_id}/tabs/{tab_id}", response_model=RequestHubTabRead)
async def update_request_hub_tab(
    workspace_id: UUID,
    hub_id: UUID,
    tab_id: UUID,
    tab_data: RequestHubTabUpdate,
    user_id: UUID = Query(..., description="User ID from auth token"),
    session: AsyncSession = Depends(get_session),
):
    """Update a request hub tab."""
    # Verify hub exists and belongs to workspace
    stmt = select(RequestHub).where(
        RequestHub.id == hub_id,
        RequestHub.workspace_id == workspace_id
    )
    result = await session.execute(stmt)
    hub = result.scalar_one_or_none()
    
    if not hub:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Request hub {hub_id} not found in workspace {workspace_id}"
        )
    
    # Get existing tab
    stmt = select(RequestHubTab).where(
        RequestHubTab.id == tab_id,
        RequestHubTab.hub_id == hub_id
    )
    result = await session.execute(stmt)
    tab = result.scalar_one_or_none()
    
    if not tab:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Tab {tab_id} not found in hub {hub_id}"
        )
    
    # Check for slug conflicts if updating slug
    if tab_data.slug and tab_data.slug != tab.slug:
        stmt = select(RequestHubTab).where(
            RequestHubTab.hub_id == hub_id,
            RequestHubTab.slug == tab_data.slug,
            RequestHubTab.id != tab_id
        )
        result = await session.execute(stmt)
        existing_tab = result.scalar_one_or_none()
        
        if existing_tab:
            raise HTTPException(
                status_code=status.HTTP_409_CONFLICT,
                detail=f"Tab with slug '{tab_data.slug}' already exists in this hub"
            )
    
    # Update fields
    update_data = tab_data.model_dump(exclude_unset=True)
    for field, value in update_data.items():
        setattr(tab, field, value)
    
    await session.commit()
    await session.refresh(tab)
    
    return tab


@router.delete("/workspaces/{workspace_id}/request-hubs/{hub_id}/tabs/{tab_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_request_hub_tab(
    workspace_id: UUID,
    hub_id: UUID,
    tab_id: UUID,
    user_id: UUID = Query(..., description="User ID from auth token"),
    session: AsyncSession = Depends(get_session),
):
    """Delete a request hub tab."""
    # Verify hub exists and belongs to workspace
    stmt = select(RequestHub).where(
        RequestHub.id == hub_id,
        RequestHub.workspace_id == workspace_id
    )
    result = await session.execute(stmt)
    hub = result.scalar_one_or_none()
    
    if not hub:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Request hub {hub_id} not found in workspace {workspace_id}"
        )
    
    # Get tab
    stmt = select(RequestHubTab).where(
        RequestHubTab.id == tab_id,
        RequestHubTab.hub_id == hub_id
    )
    result = await session.execute(stmt)
    tab = result.scalar_one_or_none()
    
    if not tab:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Tab {tab_id} not found in hub {hub_id}"
        )
    
    await session.delete(tab)
    await session.commit()


@router.post("/workspaces/{workspace_id}/request-hubs/{hub_id}/tabs/reorder", response_model=List[RequestHubTabRead])
async def reorder_request_hub_tabs(
    workspace_id: UUID,
    hub_id: UUID,
    reorder_data: TabsReorderRequest,
    user_id: UUID = Query(..., description="User ID from auth token"),
    session: AsyncSession = Depends(get_session),
):
    """Reorder tabs in a request hub."""
    # Verify hub exists and belongs to workspace
    stmt = select(RequestHub).where(
        RequestHub.id == hub_id,
        RequestHub.workspace_id == workspace_id
    )
    result = await session.execute(stmt)
    hub = result.scalar_one_or_none()
    
    if not hub:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Request hub {hub_id} not found in workspace {workspace_id}"
        )
    
    # Update positions
    for item in reorder_data.tabs:
        stmt = (
            update(RequestHubTab)
            .where(
                RequestHubTab.id == item.id,
                RequestHubTab.hub_id == hub_id
            )
            .values(position=item.position)
        )
        await session.execute(stmt)
    
    await session.commit()
    
    # Return updated tabs
    stmt = (
        select(RequestHubTab)
        .where(RequestHubTab.hub_id == hub_id)
        .order_by(RequestHubTab.position)
    )
    result = await session.execute(stmt)
    tabs = result.scalars().all()
    
    return tabs

File: go-backend/handlers/request_hubs.go
package handlers

import (
	"net/http"

	"github.com/Jsanchez767/matic-platform/database"
	"github.com/Jsanchez767/matic-platform/models"
	"github.com/gin-gonic/gin"
	"github.com/google/uuid"
)

// Request Hub Handlers

func ListRequestHubs(c *gin.Context) {
	workspaceID := c.Query("workspace_id")
	if workspaceID == "" {
		c.JSON(http.StatusBadRequest, gin.H{"error": "workspace_id is required"})
		return
	}
	includeInactive := c.Query("include_inactive") == "true"

	var hubs []models.RequestHub
	query := database.DB.Where("workspace_id = ?", workspaceID)

	if !includeInactive {
		query = query.Where("is_active = ?", true)
	}

	if err := query.Order("created_at DESC").Find(&hubs).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, hubs)
}

func GetRequestHub(c *gin.Context) {
	hubID := c.Param("hub_id")

	var hub models.RequestHub
	if err := database.DB.Preload("Tabs").
		Where("id = ?", hubID).
		First(&hub).Error; err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "Request hub not found"})
		return
	}

	c.JSON(http.StatusOK, hub)
}

func GetRequestHubBySlug(c *gin.Context) {
	workspaceID := c.Query("workspace_id")
	if workspaceID == "" {
		c.JSON(http.StatusBadRequest, gin.H{"error": "workspace_id is required"})
		return
	}
	slug := c.Param("slug")

	var hub models.RequestHub
	if err := database.DB.Preload("Tabs").
		Where("slug = ? AND workspace_id = ?", slug, workspaceID).
		First(&hub).Error; err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "Request hub not found"})
		return
	}

	c.JSON(http.StatusOK, hub)
}

type CreateRequestHubInput struct {
	WorkspaceID uuid.UUID              `json:"workspace_id" binding:"required"`
	Name        string                 `json:"name" binding:"required"`
	Slug        string                 `json:"slug" binding:"required"`
	Description string                 `json:"description"`
	Settings    map[string]interface{} `json:"settings"`
	IsActive    *bool                  `json:"is_active"`
}

func CreateRequestHub(c *gin.Context) {
	var input CreateRequestHubInput
	if err := c.ShouldBindJSON(&input); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	// Check for duplicate slug
	var existing models.RequestHub
	if err := database.DB.Where("workspace_id = ? AND slug = ?", input.WorkspaceID, input.Slug).First(&existing).Error; err == nil {
		c.JSON(http.StatusConflict, gin.H{"error": "Request hub with this slug already exists"})
		return
	}

	isActive := true
	if input.IsActive != nil {
		isActive = *input.IsActive
	}

	hub := models.RequestHub{
		WorkspaceID: input.WorkspaceID,
		Name:        input.Name,
		Slug:        input.Slug,
		Description: input.Description,
		Settings:    input.Settings,
		IsActive:    isActive,
		CreatedBy:   uuid.MustParse(c.Query("user_id")), // Get from auth middleware
	}

	if err := database.DB.Create(&hub).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	// Reload with tabs
	database.DB.Preload("Tabs").First(&hub, hub.ID)

	c.JSON(http.StatusCreated, hub)
}

type UpdateRequestHubInput struct {
	Name        *string                 `json:"name"`
	Slug        *string                 `json:"slug"`
	Description *string                 `json:"description"`
	Settings    *map[string]interface{} `json:"settings"`
	IsActive    *bool                   `json:"is_active"`
}

func UpdateRequestHub(c *gin.Context) {
	hubID := c.Param("hub_id")

	var hub models.RequestHub
	if err := database.DB.Where("id = ?", hubID).First(&hub).Error; err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "Request hub not found"})
		return
	}

	var input UpdateRequestHubInput
	if err := c.ShouldBindJSON(&input); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	// Check slug conflicts if updating
	if input.Slug != nil && *input.Slug != hub.Slug {
		var existing models.RequestHub
		if err := database.DB.Where("workspace_id = ? AND slug = ? AND id != ?", hub.WorkspaceID, *input.Slug, hubID).First(&existing).Error; err == nil {
			c.JSON(http.StatusConflict, gin.H{"error": "Request hub with this slug already exists"})
			return
		}
	}

	// Update fields
	if input.Name != nil {
		hub.Name = *input.Name
	}
	if input.Slug != nil {
		hub.Slug = *input.Slug
	}
	if input.Description != nil {
		hub.Description = *input.Description
	}
	if input.Settings != nil {
		hub.Settings = *input.Settings
	}
	if input.IsActive != nil {
		hub.IsActive = *input.IsActive
	}

	if err := database.DB.Save(&hub).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	// Reload with tabs
	database.DB.Preload("Tabs").First(&hub, hub.ID)

	c.JSON(http.StatusOK, hub)
}

func DeleteRequestHub(c *gin.Context) {
	hubID := c.Param("hub_id")

	var hub models.RequestHub
	if err := database.DB.Where("id = ?", hubID).First(&hub).Error; err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "Request hub not found"})
		return
	}

	if err := database.DB.Delete(&hub).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.Status(http.StatusNoContent)
}

// Request Hub Tab Handlers

func ListRequestHubTabs(c *gin.Context) {
	hubID := c.Param("hub_id")
	includeHidden := c.Query("include_hidden") == "true"

	// Verify hub exists
	var hub models.RequestHub
	if err := database.DB.Where("id = ?", hubID).First(&hub).Error; err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "Request hub not found"})
		return
	}

	var tabs []models.RequestHubTab
	query := database.DB.Where("hub_id = ?", hubID)

	if !includeHidden {
		query = query.Where("is_visible = ?", true)
	}

	if err := query.Order("position ASC").Find(&tabs).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, tabs)
}

type CreateRequestHubTabInput struct {
	Name      string                 `json:"name" binding:"required"`
	Slug      string                 `json:"slug" binding:"required"`
	Type      string                 `json:"type" binding:"required"`
	Icon      string                 `json:"icon"`
	Position  int                    `json:"position"`
	IsVisible *bool                  `json:"is_visible"`
	Config    map[string]interface{} `json:"config"`
}

func CreateRequestHubTab(c *gin.Context) {
	hubID := c.Param("hub_id")

	// Verify hub exists
	var hub models.RequestHub
	if err := database.DB.Where("id = ?", hubID).First(&hub).Error; err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "Request hub not found"})
		return
	}

	var input CreateRequestHubTabInput
	if err := c.ShouldBindJSON(&input); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	// Check for duplicate slug
	var existing models.RequestHubTab
	if err := database.DB.Where("hub_id = ? AND slug = ?", hubID, input.Slug).First(&existing).Error; err == nil {
		c.JSON(http.StatusConflict, gin.H{"error": "Tab with this slug already exists"})
		return
	}

	isVisible := true
	if input.IsVisible != nil {
		isVisible = *input.IsVisible
	}

	tab := models.RequestHubTab{
		HubID:     uuid.MustParse(hubID),
		Name:      input.Name,
		Slug:      input.Slug,
		Type:      input.Type,
		Icon:      input.Icon,
		Position:  input.Position,
		IsVisible: isVisible,
		Config:    input.Config,
	}

	if err := database.DB.Create(&tab).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusCreated, tab)
}

type UpdateRequestHubTabInput struct {
	Name      *string                 `json:"name"`
	Slug      *string                 `json:"slug"`
	Type      *string                 `json:"type"`
	Icon      *string                 `json:"icon"`
	Position  *int                    `json:"position"`
	IsVisible *bool                   `json:"is_visible"`
	Config    *map[string]interface{} `json:"config"`
}

func UpdateRequestHubTab(c *gin.Context) {
	hubID := c.Param("hub_id")
	tabID := c.Param("tab_id")

	// Verify hub exists
	var hub models.RequestHub
	if err := database.DB.Where("id = ?", hubID).First(&hub).Error; err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "Request hub not found"})
		return
	}

	var tab models.RequestHubTab
	if err := database.DB.Where("id = ? AND hub_id = ?", tabID, hubID).First(&tab).Error; err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "Tab not found"})
		return
	}

	var input UpdateRequestHubTabInput
	if err := c.ShouldBindJSON(&input); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	// Check slug conflicts if updating
	if input.Slug != nil && *input.Slug != tab.Slug {
		var existing models.RequestHubTab
		if err := database.DB.Where("hub_id = ? AND slug = ? AND id != ?", hubID, *input.Slug, tabID).First(&existing).Error; err == nil {
			c.JSON(http.StatusConflict, gin.H{"error": "Tab with this slug already exists"})
			return
		}
	}

	// Update fields
	if input.Name != nil {
		tab.Name = *input.Name
	}
	if input.Slug != nil {
		tab.Slug = *input.Slug
	}
	if input.Type != nil {
		tab.Type = *input.Type
	}
	if input.Icon != nil {
		tab.Icon = *input.Icon
	}
	if input.Position != nil {
		tab.Position = *input.Position
	}
	if input.IsVisible != nil {
		tab.IsVisible = *input.IsVisible
	}
	if input.Config != nil {
		tab.Config = *input.Config
	}

	if err := database.DB.Save(&tab).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, tab)
}

func DeleteRequestHubTab(c *gin.Context) {
	hubID := c.Param("hub_id")
	tabID := c.Param("tab_id")

	// Verify hub exists
	var hub models.RequestHub
	if err := database.DB.Where("id = ?", hubID).First(&hub).Error; err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "Request hub not found"})
		return
	}

	var tab models.RequestHubTab
	if err := database.DB.Where("id = ? AND hub_id = ?", tabID, hubID).First(&tab).Error; err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "Tab not found"})
		return
	}

	if err := database.DB.Delete(&tab).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.Status(http.StatusNoContent)
}

type ReorderTabInput struct {
	ID       uuid.UUID `json:"id" binding:"required"`
	Position int       `json:"position" binding:"required"`
}

type ReorderTabsInput struct {
	Tabs []ReorderTabInput `json:"tabs" binding:"required"`
}

func ReorderRequestHubTabs(c *gin.Context) {
	hubID := c.Param("hub_id")

	// Verify hub exists
	var hub models.RequestHub
	if err := database.DB.Where("id = ?", hubID).First(&hub).Error; err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "Request hub not found"})
		return
	}

	var input ReorderTabsInput
	if err := c.ShouldBindJSON(&input); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	// Update positions in transaction
	tx := database.DB.Begin()
	for _, item := range input.Tabs {
		if err := tx.Model(&models.RequestHubTab{}).
			Where("id = ? AND hub_id = ?", item.ID, hubID).
			Update("position", item.Position).Error; err != nil {
			tx.Rollback()
			c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
			return
		}
	}
	tx.Commit()

	// Return updated tabs
	var tabs []models.RequestHubTab
	if err := database.DB.Where("hub_id = ?", hubID).Order("position ASC").Find(&tabs).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, tabs)
}
